src/front/wgsl/construction.rs

```diff
@@ -184,12 +184,15 @@ fn parse_constructor_type<'a>(
             Ok(Some(ConstructorType::Vector { size, kind, width }))
         }
         (Token::Paren('<'), ConstructorType::PartialMatrix { columns, rows }) => {
-            let (_, width) = lexer.next_scalar_generic()?;
-            Ok(Some(ConstructorType::Matrix {
-                columns,
-                rows,
-                width,
-            }))
+            let (kind, width, span) = lexer.next_scalar_generic_with_span()?;
+            match (kind, width) {
+                (ScalarKind::Float, 4) => Ok(Some(ConstructorType::Matrix {
+                    columns,
+                    rows,
+                    width,
+                })),
+                _ => Err(Error::BadMatrixScalarKind(span, kind, width)),
+            }
```

Comment:
Can we use the new `parse_matrix_scalar_type` fn here as well?

Reply:
We don't have a great way to do this currently, because this returns `ConstructorType::Matrix` but the `parse_matrix_scalar_type ` returns `TypeInner::Matrix`.

We could consider making those share a type, e.g. change the enum variants to be `ConstructorType::Matrix(Matrix)` and `TypeInner::Matrix(Matrix)`. That change would affect a bunch of areas though.

Reply:
Ah right, it's ok - can leave it like this for now.